File: fries-leetcode-training/04-stack/02-min-stack.cpp
/**
 * LeetCode 155. 最小栈
 * https://leetcode.cn/problems/min-stack/
 * 
 * 设计一个支持push，pop，top操作，并能在常数时间内检索到最小元素的栈。
 * 
 * 辅助栈
 * 
 * 时间复杂度：O(1) 所有操作
 * 空间复杂度：O(n)
 */

#include <stack>
using namespace std;

class MinStack {
private:
    stack<int> data_stack;
    stack<int> min_stack;
    
public:
    MinStack() {}
    
    void push(int val) {
        data_stack.push(val);
        if (min_stack.empty() || val <= min_stack.top()) {
            min_stack.push(val);
        }
    }
    
    void pop() {
        if (data_stack.empty()) {
            return;
        }
        
        int val = data_stack.top();
        data_stack.pop();
        
        if (!min_stack.empty() && val == min_stack.top()) {
            min_stack.pop();
        }
    }
    
    int top() {
        return data_stack.top();
    }
    
    int getMin() {
        return min_stack.top();
    }
};

// 测试函数
#include <iostream>
void testMinStack() {
    MinStack minStack;
    
    minStack.push(-2);
    minStack.push(0);
    minStack.push(-3);
    
    cout << "最小元素: " << minStack.getMin() << endl;  // 期望: -3
    
    minStack.pop();
    cout << "栈顶元素: " << minStack.top() << endl;      // 期望: 0
    cout << "最小元素: " << minStack.getMin() << endl;  // 期望: -2
}

int main() {
    testMinStack();
    return 0;
}

File: fries-leetcode-training/04-stack/02-min-stack.py
"""
LeetCode 155. 最小栈
https://leetcode.cn/problems/min-stack/

设计一个支持push，pop，top操作，并能在常数时间内检索到最小元素的栈。

辅助栈

时间复杂度：O(1) 所有操作
空间复杂度：O(n)
"""

class MinStack:
    """
    最小栈 - 辅助栈法
    """
    
    def __init__(self):
        self.stack = []
        self.min_stack = []
    
    def push(self, val):
        """
        入栈
        
        Args:
            val: 要入栈的值
        """
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)
    
    def pop(self):
        """
        出栈
        """
        if not self.stack:
            return
        
        val = self.stack.pop()
        if self.min_stack and val == self.min_stack[-1]:
            self.min_stack.pop()
    
    def top(self):
        """
        获取栈顶元素
        
        Returns:
            栈顶元素
        """
        return self.stack[-1] if self.stack else None
    
    def get_min(self):
        """
        获取最小元素
        
        Returns:
            最小元素
        """
        return self.min_stack[-1] if self.min_stack else None


def test_min_stack():
    """测试函数"""
    min_stack = MinStack()
    
    # 测试用例
    min_stack.push(-2)
    min_stack.push(0)
    min_stack.push(-3)
    
    print(f"最小元素: {min_stack.get_min()}")  # 期望: -3
    
    min_stack.pop()
    print(f"栈顶元素: {min_stack.top()}")      # 期望: 0
    print(f"最小元素: {min_stack.get_min()}")  # 期望: -2


if __name__ == "__main__":
    test_min_stack()
